### PR TITLE
Renamed macros to avoid leading underscore

### PR DESCRIPTION
**Pull request by francesco-romano**
Please take a look at it.
Regex did most of the work, but a manual check is also needed (I did it, but maybe I missed something)

@drdanz  @lornat75 @traversaro

### CODE REVIEW

File: src/libYARP_dev/include/yarp/dev/PolyDriverList.h [hunk 1]
```diff
@@ -1,7 +1,7 @@
 // -*- mode:C++; tab-width:4; c-basic-offset:4; indent-tabs-mode:nil -*-
 
-#ifndef __POLYDRIVER_LIST
-#define __POLYDRIVER_LIST
+#ifndef _POLYDRIVERLIST
+#define _POLYDRIVERLIST
```
Comment (traversaro) [inline src/libYARP_dev/include/yarp/dev/PolyDriverList.h]: Something went wrong here?

**Comment by traversaro** [comment]
Not a blocking issue, but it would also be nice to fix the endif that contain a comment with the guard macros, as in https://github.com/robotology/yarp/blob/83466b733d89d2f3687542b047bcd5185d3f9299/src/libYARP_manager/include/yarp/manager/localbroker.h#L127 .

**Comment by drdanz** [comment]
:+1: to @traversaro comment... There are several headers that have the comment in the `#endif` and that is now wrong.

Seeing the patch really hurts me because we have a huge amount of different conventions... Just to make a few examples:

* `YARP2_XXX`
* `YARP_XXX`
* `YARP2XXX`
*  `YARPXXX`
* `XXX`


Also something went wrong here, or did you remove the underscore in the middle on purpose?

```diff
-#ifndef __POLYDRIVER_LIST
-#define __POLYDRIVER_LIST
+#ifndef POLYDRIVERLIST
+#define POLYDRIVERLIST
```

**Comment by francesco-romano** [comment]
No.. sometime the regex failed!
If you find some errors I will fix it.

**Comment by drdanz** [comment]
Please have a look at ``src/libYARP_sig/include/yarp/sig/Image.h``, it is changing a lot more than header guards.

I'd really want at least to avoid to have header guards in the form `XXX` in the main libraries, as they might cause unexpected issues hard to debug. Here is a list of the files:

* src/libYARP_OS/include/yarp/os/Terminator.h
* src/libYARP_OS/include/yarp/os/impl/RunProcManager.h
* src/libYARP_OS/include/yarp/os/impl/RunReadWrite.h
* src/libYARP_OS/include/yarp/os/impl/ShmemHybridStream.h
* src/libYARP_OS/include/yarp/os/impl/ShmemInputStream.h
* src/libYARP_OS/include/yarp/os/impl/ShmemOutputStream.h
* src/libYARP_OS/include/yarp/os/impl/ShmemTypes.h
* src/libYARP_OS/include/yarp/os/impl/hmac/hmac_sha2.h
* src/libYARP_dev/include/yarp/dev/ControlBoardHelpers.h
* src/libYARP_dev/include/yarp/dev/ControlBoardInterfacesImpl.h
* src/libYARP_dev/include/yarp/dev/PolyDriverDescriptor.h
* src/libYARP_dev/include/yarp/dev/PolyDriverList.h
* src/libYARP_dev/include/yarp/dev/PreciselyTimed.h
* src/libYARP_dev/src/modules/ControlBoardWrapper/ControlBoardWrapper.h
* src/libYARP_dev/src/modules/ControlBoardWrapper/RPCMessagesParser.h
* src/libYARP_dev/src/modules/ControlBoardWrapper/StreamingMessagesParser.h
* src/libYARP_dev/src/modules/ControlBoardWrapper/SubDevice.h
* src/libYARP_dev/src/modules/Rangefinder2DClient/Rangefinder2DClient.h
* src/libYARP_dev/src/modules/Rangefinder2DWrapper/Rangefinder2DWrapper.h
* src/libYARP_dev/src/modules/VirtualAnalogWrapper/VirtualAnalogWrapper.h

**Comment by francesco-romano** [comment]
@drdanz 
about `Image.h`: yes. This is why I called the PR "Renamed macros" instead of "Renamed header guards". The reserved identifier issue is on all the macros.
Anyway that macro is used only there (it is then `#undef`) so I don't see any problem

I did not understood the second point

**Comment by drdanz** [comment]
**WARNING**: In `src/modules/ffmpeg/avpreamble.h` and `src/modules/ffmpeg/ffmpeg_api.h`

```diff
-#ifndef __STDC_CONSTANT_MACROS
-#define __STDC_CONSTANT_MACROS
+#ifndef STDC_CONSTANT_MACROS
+#define STDC_CONSTANT_MACROS
```

`__STDC_CONSTANT_MACROS` is not a header guard, but a reserved macro, please change it back. A better way to do it would be to define it in the `CMakeLists.txt`, but it does not belong to this patch...

**Comment by drdanz** [comment]
@francesco-romano the second point is a bit tricky. I don't remember the details, but I had once to debug for hours an issue that was caused by an enum having as a value something that was defined by a header guard of a library included. I can guarantee that when I found the reason, I cursed the developers for using a "generic" name. ``#define TERMINATOR`` is something that reminds me a lot about that.

**Comment by traversaro** [comment]
I did not notice tha GitHub was hidings some modified files... I still have to review the real diff then.

**Comment by drdanz** [comment]
@francesco-romano About `Image.h`, do you mind splitting it in a separate commit? It will make it easier to bisect in case something goes wrong...

**Comment by francesco-romano** [comment]
With @drdanz and @lornat75 we decided the following format for the header guards:

`YARP_[libname]_[filename]_H` all capital

**Comment by drdanz** [comment]
Expanding a little, I propose:

* Main libraries: `YARP_<LIB>_<FILENAME_H>` i.e. `YARP_OS_FOO_H`
* Impl headers: `YARP_<LIB>_IMPL_<FILENAME_H>` i.e. `YARP_OS_IMPL_FOO_H`
* Plugins (carriers and modules): `YARP_<NAME>_<CARRIER/DEVICE>_FILENAME_H` i.e. `YARP_HUMAN_CARRIER_HUMANSTREAM_H` and `YARP_FAKEBOT_DEVICE_FAKEBOT_H`
* Executables (tools and GUIs): `YARP_<NAME>_FILENAME_H` i.e. `YARP_YARP_YARPROBOT_H` (:scream: this one is ugly on purpose) and `YARP_YARPMANAGER_QT_YSCOPEWINDOW_H`

We can probably forget about plugins and executable though, since they are never included outside their code.

**Comment by francesco-romano** [comment]
@drdanz I've just seen your comment. I'll further modify the files later. 
Library files should be ok though.

**Comment by drdanz** [comment]
@francesco-romano It's just a proposal, if you don't like it, we can discuss about that...

**Comment by francesco-romano** [comment]
@drdanz lib and impl done.
I do not understand the modules/plugin.
How would you do for example the `src/libYARP_dev/src/modules/ControlBoardWrapper/ControlBoardWrapper.h`?
Right now it is `YARP_DEV_CBW_CONTROLBOARDWRAPPER_H`.

**Comment by drdanz** [comment]
@francesco-romano I don't know... :smile:
Perhaps `YARP_DEV_CONTROLBOARDWRAPPER_CONTROLBOARDWRAPPER_H`?

**Comment by barbalberto** [comment]
I suggest ```YARP_DEV_MODULE_CONTROLBOARDWRAPPER2_H``` ?

**Comment by francesco-romano** [comment]
OK.
Maybe we can take a decision for the style of modules and binaries before further modifying the code?

**Comment by drdanz** [comment]
It's probably not going to happen, but assuming that 2 different modules have some file with the same name it might conflict.

Since they are modules in dev, perhaps

`YARP_DEV_CONTROLBOARDWRAPPER_MODULE_CONTROLBOARDWRAPPER_H`

[Google style guide](https://google.github.io/styleguide/cppguide.html#The__define_Guard) suggests to use the full path, but I think that's too long for us, they also suggest to add an `_` at the end, but if I understood correctly this causes warnings on OSX, right?

**Comment by francesco-romano** [comment]
I agree with @drdanz about the name clashing.

Usually I don't like google style guide. Furthermore, even if the warning is for starting underscore, adding a trailing underscore breaks symmetry :stuck_out_tongue: 

I vote for the previous proposal by @drdanz , i.e., `YARP_DEV_CONTROLBOARDWRAPPER_CONTROLBOARDWRAPPER_H`. The reason is: adding module is not necessary since we are already deviating from the library rule.

**Comment by francesco-romano** [comment]
@drdanz done.
Can you take a look at it now, and merge if it is ok?

**Comment by drdanz** [comment]
LGTM now.
Merging.

**Comment by barbalberto** [comment]
Can a page on the wiki be published explaing the policy for people creating new files?
And send an email to yarp guys advertising it?

**Comment by traversaro** [comment]
@barbalberto Good idea, but I would avoid the wiki and document everything related to YARP in yarp documentation (a "modifyng YARP faqs" document in https://github.com/robotology/yarp/tree/master/doc would be ideal).

**Comment by drdanz** [comment]
@francesco-romano @barbalberto @traversaro Added the information in the [CONTRIBUTING.md](https://github.com/robotology/yarp/blob/master/.github/CONTRIBUTING.md) file (work in progress).

**Pull request merged**